### ISSUE TITLE
client local model 更新方法貌似有误

### ISSUE THREAD

**Issue by KarhouTam** [ki-ljl/Per-FedAvg#1]
我看你的 `client.train()` 里每个 local epoch 只是单纯走了两个 `one_step()`， 并不涉及 paper 里的二阶梯度计算。

**Comment by ki-ljl** [comment]
我们观察Per-FedAvg的优化函数：
![image](https://user-images.githubusercontent.com/56509367/163300571-e2c17b8f-9dac-4324-bc7b-e9f0a24f69fb.png)
该优化函数和FedAvg相比仅有一点不同：**FedAvg直接优化原始Server参数对应的损失函数，而Per-FedAvg优化经过一步梯度下降后得到的Server参数对应的损失函数。**

Per-FedAvg与FedAvg的另一点不同在于：**FedAvg每次都需要对所有bacth进行训练，但Per-FedAvg只选取一批。**

我们可以首先进行第一个`model = one_step()`，此时model中的参数将变成meta function中的参数：
![image](https://user-images.githubusercontent.com/56509367/163301278-b9bc826f-9b2e-4a3e-85f1-264c5fc265a4.png)
代码表示为：
```python
def one_step_1(args, data, model, lr):
    ind = np.random.randint(0, high=len(data), size=None, dtype=int)
    seq, label = data[ind]
    seq = seq.to(args.device)
    label = label.to(args.device)
    y_pred = model(seq)
    optimizer = torch.optim.Adam(model.parameters(), lr=lr)
    loss_function = nn.MSELoss().to(args.device)
    loss = loss_function(y_pred, label)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    return model

model = one_step_1(args, data, model, lr=args.alpha)
```
然后再进行第二个`one_step()`，得到meta function的梯度，然后对**原始参数**进行更新，而不是meta function中的参数，这里有一些错误，正确做法应该是：
```python
def one_step_2(args, data, original_model, model, lr):
    ind = np.random.randint(0, high=len(data), size=None, dtype=int)
    seq, label = data[ind]
    seq = seq.to(args.device)
    label = label.to(args.device)
    # meta function parameter
    y_pred = model(seq)
    # update original model
    optimizer = torch.optim.Adam(original_model.parameters(), lr=lr)
    loss_function = nn.MSELoss().to(args.device)
    loss = loss_function(y_pred, label)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()

    return original_model

model = one_step_2(args, data, origin_model, model, lr=args.beta)
```
`one_step_2()`中包含了求二阶梯度。

完整代码：
```python
def train(args, model):
    model.train()
    original_model = copy.deep_copy(model)
    Dtr, Dte = nn_seq_wind(model.name, args.B)
    model.len = len(Dtr)
    print('training...')
    data = [x for x in iter(Dtr)]
    for epoch in range(args.E):
        # step1
        model = one_step_1(args, data, model, lr=args.alpha)
        # step2
        model = one_step_2(args, data, original_model, model, lr=args.beta)

    return model
```

**Comment by KarhouTam** [comment]
我看你贴的 `one_step_2()` 代码
```python    
y_pred = model(seq)
# update original model
optimizer = torch.optim.Adam(original_model.parameters(), lr=lr)
loss_function = nn.MSELoss().to(args.device)
loss = loss_function(y_pred, label)
optimizer.zero_grad()
loss.backward()
optimizer.step()
```
你的 `pred_y` 是用 `model` 计算的, 而你的 `optimizer` 优化的参数组是 `original_model.parameters()` 。但你的 `original_model = deepcopy(model)`，是个完全不同的对象。用 `model` 来计算 `loss`，pytorch 并不会生成跟你的 `original_model` 相关的计算图， 你的 `optimizer.step()` 其实并不会真的优化到 `original_model.parameters()`。 且 paper 里的公式
![屏幕截图(1)](https://user-images.githubusercontent.com/82705454/163324884-710af433-8aa8-48d5-b21f-23d0f56217d9.png)
我的理解是先 deepcopy 一个 model 得到 replica， 用一个 data batch 在 replica 做一次 gd，再用这个 replica 求出一个梯度， 同时用 original model 来求二阶梯度。 你说的 `one_step_1()`，`one_step_2()` 本质上应该是对一个模型做两次 gd。如果我说的有错，请指正。

**Comment by ki-ljl** [comment]
> 我看你贴的 `one_step_2()` 代码
> 
> ```python
> y_pred = model(seq)
> # update original model
> optimizer = torch.optim.Adam(original_model.parameters(), lr=lr)
> loss_function = nn.MSELoss().to(args.device)
> loss = loss_function(y_pred, label)
> optimizer.zero_grad()
> loss.backward()
> optimizer.step()
> ```
> 
> 你的 `pred_y` 是用 `model` 计算的, 而你的 `optimizer` 优化的参数组是 `original_model.parameters()` 。但你的 `original_model = deepcopy(model)`，是个完全不同的对象。用 `model` 来计算 `loss`，pytorch 并不会生成跟你的 `original_model` 相关的计算图， 你的 `optimizer.step()` 其实并不会真的优化到 `original_model.parameters()`。 且 paper 里的公式 ![屏幕截图(1)](https://user-images.githubusercontent.com/82705454/163324884-710af433-8aa8-48d5-b21f-23d0f56217d9.png) 我的理解是先 deepcopy 一个 model 得到 replica， 用一个 data batch 在 replica 做一次 gd，再用这个 replica 求出一个梯度， 同时用 original model 来求二阶梯度。 你说的 `one_step_1()`，`one_step_2()` 本质上应该是对一个模型做两次 gd。如果我说的有错，请指正。

这个思路就是我代码中注释掉的那部分，不过当时好像报错了，我就没再弄了，你可以看一下。

**Comment by KarhouTam** [comment]
其实你可以实现 paper 中说到的 Per-FedAvg(FO)，那个就是跟你说的一样的（不过那个 `one_step_2()` 的  `optimizer` 优化的参数组不对是 bug，我 debug 你的代码发现 `step()` 前后 `original_model` 参数没变动）。至于求二阶梯度，我倒是有种方法，但是实在过于麻烦，我本意其是就是想找别人的代码看看怎么实现的。😂

**Comment by KarhouTam** [comment]
至于你 comment 掉的那部分代码。 `backward()` 两次并不会产生二阶梯度，pytorch 会帮你计算两次一阶梯度。计算二阶梯度需要用 `torch.autograd.grad(grad_1st, model.parameters())` 来计算。但 `grad_1st` 并不是 scalar， 所以处理起来会很麻烦。😂
